{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/LACDR-CDS/SCDR_Bioinformatics_Practical/blob/main/Session2/Session2.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ti6ze0BesaCJ"
      },
      "source": [
        "# Session 2"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "J1iriuNIsiDf"
      },
      "source": [
        "In this session, you will go through the basic processing steps of RNA sequencing data. Using the steps taught in the lecture and repeated below, as well as the **cheat sheet** provided on Brightspace/Teams, you will write your own code to process and analyze the data.\n",
        "\n",
        "Each team receives a count matrix with 4 samples: **2 ESC replicates** and 2 replicates which belong to one of the following stages of cardiomyocyte differentiation:\n",
        "\n",
        "*   mesoderm\n",
        "*   cardiomyocyte progenitors\n",
        "*   cardiomyocytes\n",
        "*   fetal heart\n",
        "\n",
        "Your goal is to **label the samples with the correct stage** (both ESCs and differentiated).\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zUtVXsC0smtf"
      },
      "source": [
        "## Setup\n",
        "\n",
        "Run the following cells to set up the necessary packages and download the data. If you wish to use a package which is not in the list below, you will need to install and import it yourself."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "9YB4CMxas7um"
      },
      "outputs": [],
      "source": [
        "#Install packages which are not in the default environment\n",
        "%pip install scanpy\n",
        "%pip install pydeseq2\n",
        "%pip install gseapy"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "A11sAvI2OcBY"
      },
      "outputs": [],
      "source": [
        "#Import packages\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "import scanpy as sc\n",
        "import anndata as ad\n",
        "import os\n",
        "import pandas as pd\n",
        "from sklearn.preprocessing import StandardScaler\n",
        "from sklearn.decomposition import PCA\n",
        "from pydeseq2.dds import DeseqDataSet\n",
        "from pydeseq2.ds import DeseqStats\n",
        "import pickle\n",
        "import gseapy\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "group_number ="
      ],
      "metadata": {
        "id": "ENP90GLr5-Ko"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "RuXi2uAqaPQK"
      },
      "outputs": [],
      "source": [
        "#Make data directory if it does not exist\n",
        "os.makedirs(\"data\", exist_ok=True)\n",
        "os.makedirs(\"plots\", exist_ok=True)\n",
        "\n",
        "#Download datasets in the data folder\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session2/datasets/group{group_number}.txt -O data/group{group_number}.txt\n",
        "!wget https://raw.githubusercontent.com/LACDR-CDS/SCDR_Bioinformatics_Practical/refs/heads/main/Session2/datasets/metadata.csv -O data/group{group_number}_metadata.csv"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xMGaLbfJlcWD"
      },
      "source": [
        "## Data import\n",
        "\n",
        "In the empty cell(s) below, please do the following and answer the given questions:\n",
        "\n",
        "1) Read the count matrix (the path should be: `\"data/groupN.txt`\", where N is your group number).\n",
        "2) Check the orientation of the count table (`genes x samples` or `samples x genes`).\n",
        "3) How many samples and genes there are in the count table?\n",
        "4) Read the corresponding metadata table ( `\"data/groupN_metadata.csv`\"). What cell line do the samples belong to?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "EkGNpY0gN-QX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dYkg-niFn2Q0"
      },
      "source": [
        "## Filtering"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "fBVRJRkFsU26"
      },
      "source": [
        "Now that you have become familiar with the structure of your **raw data**, you can start the processing steps. First, the data must be filtered such that we remove **genes with less than 10 reads over all samples**. We do this to remove possible noise/experimental artefacts, which decrease statistical confidence. Reference the cheat sheet for help with coding. Perform the filtering and answer the following questions:\n",
        "\n",
        "1) Which genes have low overall read counts (lowest total gene count number that is higher than 0)?\n",
        "\n",
        "*Hint: use gene_count_sums.sort_values() to sort the results*\n",
        "\n",
        "2) How many genes are left in the count table after filtering?\n",
        "3) How many reads does each sample have for OCT4?\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "jOuI_IUrgx7O"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "LiuPx-KOoMAE"
      },
      "source": [
        "## Normalization: counts per million\n",
        "\n",
        "In order to account for differences in sequencing depth among samples (a sample that has been sequenced deeper will also have more gene counts overall, but this does not necessarily reflect the biology), we **normalize the gene counts** to the **total number of reads in a sample** and multiply by **1 million**. Use the cheat sheet to help you write the code. Normalize your count table and answer the following questions:\n",
        "\n",
        "1. Which sample is the most deeply sequenced and which one is the most shallow? Can you see this from the normalized table?\n",
        "2. Which gene has the highest average expression value after normalization?\n",
        "*Hint: take the average of each row and sort it in descending order using df.sort_values(ascending = False) to keep the gene names visible.*"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "Z0uyLw_ZEwXn"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_Kzuhv3spxM0"
      },
      "source": [
        "## Principal Component Analysis\n",
        "\n",
        "Principal Component Analysis (PCA) is a dimension reduction method. In order to visualise the distribution of our samples in a 2D plot, we need to reduce the count matrix from its high number of dimensions (in this case, the read count value for each gene represents a dimension) to less, out of which we then plot the first two.\n",
        "\n",
        "1) PCA requires the data to be transposed and scaled to mean 0 and unit variance. Use the cheat sheet to find the correct function and standardize the data.\n",
        "2) Run the PCA function. What is the shape of the resulting matrix? What do the rows and columns represent? (which are the samples and which are the principal components?)\n",
        "3) Plot the first two principal components. How many samples have you plotted?\n",
        "4) Which samples group together? Are there any outliers?"
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "OWHMcEzd5viV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xonJUdBRynbo"
      },
      "source": [
        "### Gene markers\n",
        "\n",
        "Now it is time to label your data. Use the cheat sheet to create a heatmap plot showing known marker genes for each possible differentiation stage:\n",
        "1. Embryonic Stem Cells\n",
        "2. Mesoderm\n",
        "3. Cardiomyocyte progenitors\n",
        "4. Cardiomyocytes\n",
        "5. Mature fetal heart\n",
        "\n",
        "Label the stem cell replicates. Then, identify which stage the other two samples belong to according to the gene expression values in your heatmap. Feel free to consult among yourselves and compare heatmaps if you find it difficult to choose between potential stages.\n",
        "\n",
        "Add your chosen labels to the metadata table (replace the 'Unknown' in column 'stage' with your chosen label) and verify the clustering in a labelled PCA plot."
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "B_P0ZZfE7h9w"
      },
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}